### PR TITLE
Correct passthrough sequences after refactor

### LINKED ISSUE #4116: ConPTY pass-through sequences are being truncated
# Environment

Windows build number: Version 10.0.18362.418
Windows Terminal version (if applicable): commit d711d731d7d00259b2e0fe5847928c9f1c1121ed

# Steps to reproduce

1. Open a bash shell in Windows Terminal
2. Execute the sequence: `printf "\e[?999h 12345 Hello World"`

# Expected behavior

999 isn't a supported mode, so that should be ignored and passed through to the conpty client. Then the text ` 12345 Hello World` should be output.

# Actual behavior

The unrecognized mode sequence gets truncated, losing the final character. and then the start of the text that is output is also truncated, leaving just `ello World`.

What's happening is that the missing character leaves the state machine in a state where it's waiting for a final character, so it ends up eating the characters that follow, until it gets to the `H` in `Hello World`.

I believe this is a regression caused by PR #3956. I think it's the calculation of the `_run` variable that is off by 1, but it may be more complicated than that:

https://github.com/microsoft/terminal/blob/d711d731d7d00259b2e0fe5847928c9f1c1121ed/src/terminal/parser/stateMachine.cpp#L1252

Also note that `_run` is initialized in more than one place.

### PR DESCRIPTION
**Pull request by miniksa**
## Summary of the Pull Request
When refactoring the `StateMachine::ProcessString` algorithm to use safer structures, I made an off-by-one error when attempting to simplify the loop. 

## References
- Introduced in #3956 

## PR Checklist
* [x] Closes #4116
* [x] I work here.
* [x] Tests added/passed
* [x] No documentation
* [x] I'm a core contributor.

## Detailed Description of the Pull Request / Additional comments
The algorithm in use exploited holding onto some pointers and sizes as it rotated around the loop to call back as member variables in the pass-through function `FlushToTerminal`. 

As a part of the refactor, I adjusted to persisting a `std::wstring_view` of the currently processing string instead of pointer/size. I also attempted to simplify the loop at the same time as both the individual and group branches were performing some redundant operations in respect to updating the "run" length. 

Turns out, I made a mistake here. I wrote it so it worked correctly for the bottom half where we transition from bulk printing to an escape but then I messed up the top case.

## Validation Steps Performed
- [x] Manual validation of the exact command given in the bug report.
- [x] Wrote automated tests to validate both paths through the `ProcessString` loop that work with the `_run` variable.

### CODE REVIEW

File: src/terminal/parser/ut_parser/StateMachineTest.cpp [hunk 1]
```diff
@@ -104,3 +134,47 @@ void StateMachineTest::TwoStateMachinesDoNotInterfereWithEachother()
...
+    auto enginePtr{ std::make_unique<TestStateMachineEngine>() };
+    // this dance is required because StateMachine presumes to take ownership of its engine.
+    auto& engine{ *enginePtr.get() };
+    StateMachine machine{ std::move(enginePtr) };
+
+    // Print a bunch of regular text to build up the run buffer before transitioning state.
+    machine.ProcessString(L"12345 Hello World\x1b[?999h");
+
+    // Then ensure the entire buffered run was printed all at once back to us.
+    VERIFY_ARE_EQUAL(String(L"12345 Hello World"), String(engine.printed.c_str()));
```
Comment (DHowett-MSFT) [inline src/terminal/parser/ut_parser/StateMachineTest.cpp]: verify passthrough is okay?

Comment (miniksa) [reply src/terminal/parser/ut_parser/StateMachineTest.cpp]: I guess it doesn't hurt. OK.

**Comment by DHowett-MSFT** [comment]
Nit: reword title to be imperative

**Comment by miniksa** [comment]
Checked x86 conparser.unit.tests.dll locally
- [x] Debug passed
- [x] Release passed

So I'm going to re-run the x86 failed job.

**Pull request merged**